### PR TITLE
Get individual order details API updated / SSS-71

### PR DESCRIPTION
**Pull request by ACC-Vaibhav-Palvi**
Added URL for get individual order details

### CODE REVIEW

File: src/main/java/com/angelbroking/smartapi/SmartConnect.java [hunk 1]
```diff
@@ -804,6 +804,22 @@ public JSONObject logout() {
...
+	 *
+	 * @return JSONObject which contains order details from Smart API
+	 *
+	 */
+	public JSONObject getIndividualOrderDetails(String orderId) {
+		try {
+			String url = routes.get("api.individual.order").concat(orderId);
+			return smartAPIRequestHandler.getRequest(this.apiKey, url, accessToken);
+		} catch (SmartAPIException e) {
+            throw new RuntimeException(e);
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/SmartConnect.java]: why catch (SmartAPIException e) {
            throw new RuntimeException(e);?

**Review by anurag-amx — CHANGES_REQUESTED** [review]
Dev Test case is missing.

File: src/main/java/com/angelbroking/smartapi/sample/Examples.java [hunk 2]
```diff
@@ -169,6 +169,7 @@ public void getHolding(SmartConnect smartConnect) throws SmartAPIException, IOEx
 	public void getAllHolding(SmartConnect smartConnect) throws SmartAPIException, IOException {
 		// Returns All Holding.
 		JSONObject response = smartConnect.getAllHolding();
+		System.out.println(response);
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/sample/Examples.java]: System.out.println(response); we should not SOP statements and use logger instead.

**Review by anurag-amx — CHANGES_REQUESTED** [review]

**Comment by ACC-Vaibhav-Palvi** [comment]
> Dev Test case is missing.

Hi @anurag-amx  , i write test case for tthis, will reflect 

> Dev Test case is missing.

@anurag-amx Have added test case for it, please check

File: src/main/java/com/angelbroking/smartapi/sample/Examples.java (hunk 2 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/sample/Examples.java]: > System.out.println(response); we should not SOP statements and use logger instead.

remove SOP from all over the project and added logger

File: src/main/java/com/angelbroking/smartapi/SmartConnect.java (hunk 1 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/SmartConnect.java]: The exceptions now handle correctly , please check

**Review by anurag-amx — CHANGES_REQUESTED** [review]
1. How to control logger (error , infor , warn debug) through property file.
2. How to change console logging to file logging.

**Comment by ACC-Vaibhav-Palvi** [comment]
> 1. How to control logger (error , infor , warn debug) through property file.
> 2. How to change console logging to file logging.


We can control logger through property file but i am using an annotation  @Slf4j, which is part of the Lombok library, is a Java annotation that simplifies the creation of logging functionality within our application. With this annotation, we can achieve the following benefits:

Reduced Boilerplate Code: Traditionally, logging in Java requires the declaration of static final logger fields in each class. This results in a considerable amount of repetitive code. By using @Slf4j, we eliminate this boilerplate, making our code cleaner and more concise.
Improved Readability: With @Slf4j, our code becomes more readable and focused on the actual business logic rather than the setup and maintenance of logger instances.
Consistency: @Slf4j ensures that logging is consistently implemented across our entire project, reducing the risk of errors and inconsistencies.
Flexibility: Although @Slf4j simplifies logger creation, it is fully compatible with various logging frameworks, allowing us to configure our logging as needed, including log levels, log destinations, and more.

File: src/main/java/com/angelbroking/smartapi/SmartConnect.java [hunk 3]
```diff
@@ -800,10 +800,32 @@ public JSONObject logout() {
...
+	 */
+	public JSONObject getIndividualOrderDetails(String orderId) throws IOException, SmartAPIException {
+		try {
+			String url = routes.get("api.individual.order").concat(orderId);
+			return smartAPIRequestHandler.getRequest(this.apiKey, url, accessToken);
+		}catch (SmartAPIException ex) {
+			log.error("{} while getting individual order {}", SMART_API_EXCEPTION_OCCURRED, ex.toString());
+			throw new SmartAPIException(String.format("%s in getting individual order %s", SMART_API_EXCEPTION_ERROR_MSG, ex));
+		} catch (IOException ex) {
+			log.error("{} while getting individual order {}", IO_EXCEPTION_OCCURRED, ex.getMessage());
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/SmartConnect.java]: Remove } catch (IOException ex) {
			log.error("{} while getting individual order {}", IO_EXCEPTION_OCCURRED, ex.getMessage());
			throw new IOException(String.format("%s in getting individual order %s", IO_EXCEPTION_ERROR_MSG, ex.getMessage()));
		} catch (JSONException ex) {
			log.error("{} while getting individual order {}", JSON_EXCEPTION_OCCURRED, ex.getMessage());
			throw new JSONException(String.format("%s in getting individual order %s", JSON_EXCEPTION_ERROR_MSG, ex.getMessage()));

		} 

They will never triggered.

**Review by anurag-amx — CHANGES_REQUESTED** [review]

File: src/main/java/com/angelbroking/smartapi/SmartConnect.java (hunk 3 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/SmartConnect.java]: These unnessary catch blocks are now removed. Now only one catch block is added with SmartAPI exception is used to handled the exception.

` @Test
    public void testIndividualOrder_Success() throws SmartAPIException, IOException {
        String url = routes.get("api.individual.order");
        when(smartAPIRequestHandler.getRequest(eq(this.apiKey), eq(url), eq(this.accessToken))).thenReturn(createIndividualOrderResponse());
        try {
            JSONObject response = smartAPIRequestHandler.getRequest(this.apiKey, url, this.accessToken);
            JSONObject data = response.getJSONObject("data");
            assertNotNull(data);
        } catch (SmartAPIException ex) {
            log.error("{} while getting individual order {}", SMART_API_EXCEPTION_OCCURRED, ex.toString());
            throw new SmartAPIException(String.format("%s in getting individual order %s", SMART_API_EXCEPTION_ERROR_MSG, ex));
        }
    }`

**Review by anurag-amx — CHANGES_REQUESTED** [review]
No code changes found as requested and also please move string to constant file.

File: src/main/java/com/angelbroking/smartapi/SmartConnect.java (hunk 3 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/SmartConnect.java]: Actually I have test these catch blocks now and i found that these catch blocks effectively handled exceptions thrown by the **handle** method when invoking  APIs with okkhttp dependency. This ensured robust error handling during API calls. 
So i have again made the changes in the code and add that previous catch blocks.

**Comment by ACC-Vaibhav-Palvi** [comment]
> No code changes found as requested and also please move string to constant file.

Moving the string to a constant file requires changes across the codebase, which might affect other code. I'll address this globally. Could you please create a new ticket to track these changes more effectively?

**Review by anurag-amx — APPROVED** [review]

**Pull request merged**